### PR TITLE
Hstore for keeping metric dimension

### PR DESCRIPTION
**Pull request by starstuck**
This significantly simplifies database schema by using postgres native hash structures for keeping and indexing dimensions. Benefit is smaller space usage and better performance.

### CODE REVIEW

File: src/command_builder.erl [hunk 1]
```diff
@@ -1,27 +1,37 @@
...
-            "ON CONFLICT DO NOTHING RETURNING id",
-    Values = [Collection, Metric, Bucket, Key],
+            "(collection, metric, bucket, key, dimensions) VALUES "
+            "($1, $2, $3, $4, $5) "
+            "ON CONFLICT DO NOTHING RETURNING dimensions",
+    HStore = tags_to_hstore(Tags),
+    Values = [Collection, Metric, Bucket, Key, HStore],
     {ok, Query, Values}.
 
+-spec delete_metric(binary(), [binary()], binary(), [binary()]) ->
```
Comment (elvisci) [inline src/command_builder.erl]: Line 33 has 1 trailing whitespace characters.

File: src/query_builder.erl [hunk 2]
```diff
@@ -21,339 +26,273 @@ collections_query() ->
...
+    OpStr = case Op of
+                '?&' -> "?&";
+                '?|' -> "?|"
+            end,
+    Cond = ["dimensions ", OpStr, " $", i2l(I+1)],
+    {Tags, _} = lists:unzip(Parts),
+    Keys = lists:map(fun encode_tag/1, Tags),
+    {Cond, [Keys]};
+%% joining operators
+criteria_condition({Op, L, R}, I)
```
Comment (elvisci) [inline src/query_builder.erl]: Line 239 has 1 trailing whitespace characters.

File: src/query_builder.erl [hunk 3]
```diff
@@ -21,339 +26,273 @@ collections_query() ->
...
+    LenCrit = {{tag, <<"ddb">>, <<"key_length">>}, integer_to_binary(Count)},
+    {'@>', [LenCrit | Acc]};
+criteria_from_glob_pattern(['*' | Rest], Count, Acc) ->
+    criteria_from_glob_pattern(Rest, Count + 1, Acc);
+criteria_from_glob_pattern([Part | Rest], Count, Acc) ->
+    CountBin = integer_to_binary(Count + 1),
+    TName = <<"part_",  CountBin/binary>>,
+    Acc1 = [{{tag, <<"ddb">>, TName}, Part} | Acc],
+    criteria_from_glob_pattern(Rest, Count + 1, Acc1).
+
```
Comment (elvisci) [inline src/query_builder.erl]: Line 284 has 4 trailing whitespace characters.

File: src/command_builder.erl [hunk 4]
```diff
@@ -35,47 +45,30 @@ delete_metric(Collection, Metric, Bucket, Key)
...
-tag_values(P)  ->
-    ["($", i2l(P), ", "
-     "$", i2l(P + 1), ", "
-     "$", i2l(P + 2), ", "
-     "$", i2l(P + 3), ", "
-     "$", i2l(P + 4), ")"].
-
-i2l(I) ->
-    integer_to_list(I).
+-spec delete_tags(binary(), [binary()], binary(), [binary()],
```
Comment (elvisci) [inline src/command_builder.erl]: Line 62 has 1 trailing whitespace characters.

File: src/query_builder.erl [hunk 5]
```diff
@@ -21,339 +26,273 @@ collections_query() ->
...
+    {'or', L1, R1}.
+%% We can do further improvements using '?&' and '?!' operators once we add
+%% support for tag presence only conditions.
+
+%% special, optimized operators
+criteria_condition({'@>', Parts}, I)  ->
+    Cond = ["dimensions @> $", i2l(I+1)],
+    HStore = {[{encode_tag(T), V} || {T, V} <- Parts]},
+    {Cond, [HStore]};
+criteria_condition({Op, Parts}, I)
```
Comment (elvisci) [inline src/query_builder.erl]: Line 228 has 1 trailing whitespace characters.

File: src/command_builder.erl [hunk 6]
```diff
@@ -35,47 +45,30 @@ delete_metric(Collection, Metric, Bucket, Key)
...
-        "DO UPDATE SET value = excluded.value",
-    build_tags(MID, Collection, 1, OnConflict, Tags, Query, []).
-
-delete_tag(MetricID, Namespace, TagName) ->
-    Query = "DELETE FROM " ?DIM_TABLE " WHERE "
-            "metric_id = $1 "
-            "AND namespace = $2 "
-            "AND name = $3",
-    Values = [MetricID, Namespace, TagName],
+-spec update_tags(binary(), [binary()], binary(), [binary()],
```
Comment (elvisci) [inline src/command_builder.erl]: Line 48 has 1 trailing whitespace characters.

File: src/query_builder.erl [hunk 7]
```diff
@@ -21,339 +26,273 @@ collections_query() ->
...
+    encode_tag_key(Ns, Name).
 
-grouping_join([]) ->
-    " ";
-grouping_join([N | R]) ->
-    ["INNER JOIN ", ?DIM_TABLE, " AS ", N,
-     " ON ", N, ".metric_id = ", ?MET_TABLE ".id " | grouping_join(R)].
+escape_sql_pattern(<<>>, Acc) ->
+    Acc;
+escape_sql_pattern(<<C:8/integer, Rest/binary>>, Acc)
```
Comment (elvisci) [inline src/query_builder.erl]: Line 293 has 1 trailing whitespace characters.

File: src/dqe_idx_pg.erl [hunk 8]
```diff
@@ -225,17 +188,12 @@ tdelta(T0) ->
...
-            ok;
-        E ->
-            E
-    end.
+decode_ns_rows(Rows) ->
+    [decode_ns(E) || {E} <- Rows].
 
 execute({select, Name, Q, Vs}) ->
     T0 = erlang:system_time(),
+    io:format("Query: ~s~n <- ~p~n", [Q, Vs]),
```
Comment (elvisci) [inline src/dqe_idx_pg.erl]: Remove the debug call to io:format/2 on line 196.

**Comment by Licenser** [comment]
@davecromberge added some really nice EQC tests a while back that check most of the functionality. They do seem to fail all over the place now. I suspect partially because of internal API changes but it'd be really nice if we could carry them over.

**Comment by starstuck** [comment]
Which tests do you mean. Running `make eqc-test` succeeded when I was testing this. Are there any other test? I am happy to carry those over, just tell me where I can find them? Is it on separate branch?

**Comment by Licenser** [comment]
interesting let me re-run it exploded everywhere for me.

**Comment by starstuck** [comment]
Maybe test db had some data in old schema.

**Comment by Licenser** [comment]
You are absolutely right, I just needed to delete the _build folder, got to remember that :) sorry!

**Comment by Licenser** [comment]
You said you had migrating data sorted already? Could we add SQL file for that?

**Comment by starstuck** [comment]
Sure, but where? Just to readme?

**Comment by starstuck** [comment]
It is rather expensive, because it has to re-create a lot of indexes, so I think for most people it will require manual intervention any way?

**Comment by Licenser** [comment]
Think README is fine, or a .SQL in `priv` just to have it so we're not influx-dbing things ;)

**Comment by starstuck** [comment]
OK, let me test instructions and I will update README with them. Is it fine for you to have 2 step process, which only requires to switch processes writing for a time of filling dimensions column?

**Comment by Licenser** [comment]
It's totally fine to have a 10 step process :) just want to prevent that we have to tell people "0.4.0 will delete all your indexes"

**Pull request merged**